### DISCUSSION TITLE
Test discovery should skip classes with protected constructors

### DISCUSSION THREAD

**Discussion by dasblinkenlight** [xunit/xunit#2781] [Ideas]
I am unit-testing code against different SQL databases (MySql and SQLite)
I created a test class that receives a `DbTestFixture` object in the constructor, and uses the DB connection from it inside its `[Fact]` methods, like this:

```
public class DbTest : IDisposable {
    private readonly DbConnection db;
    protected DbTest(DbTestFixture fixture) {
        db = fixture.Db;
    }
    [Fact]
    public Create() { ... }
    [Fact]
    public Update() { ... }
    [Fact]
    public Delete() { ... }
}
```
Then I created an abstract `DbTestFixture` class, and made two implementations of it - one for each database, like this:
```
public abstract class DbTestFixture : IDisposable {
    public DbConnection Db { get; protected set; }
}
public class MySqlTestFixture : DbTestFixture { ... }
public class SqliteTestFixture : DbTestFixture { ... }
```
Finally, I defined subclasses of `DbTest` also inheriting `IClassFixture<T>` for the corresponding test fixture implementation, like this:
```
public class MySqlTest : DbTest, IClassFixture<MySqlTestFixture> {
    public MySqlTest(MySqlTestFixture fixture) : base(fixture) {
    }
}
public class SqliteTest : DbTest, IClassFixture<SqliteTestFixture> {
    public SqliteTest(SqliteTestFixture fixture) : base(fixture) {
    }
}
```
This setup works almost perfectly, except for one thing: my `DbTest` class gets discovered along with other tests, and fails all tests because the framework cannot construct an instance since the constructor is protected.

It would be great if the test discovery code skipped classes with protected constructors automatically. What do you think?

Note: I am open to suggestions of a better way to do what I am trying to do, without providing my own `[assembly: TestFramework...]` implementation.

**Comment by bradwilson** [comment]
I believe this is working as I'd expect it to: explicit failure is better than silent failure.

A non-abstract class with test methods in it is something the test framework considers to be executable. Skipping the class because it has an issue is not really the way we operate. It's too clever, and "clever" is much worse than "clear".

Any reason you don't mark `DbTest` as `abstract`?

**Comment by dasblinkenlight** [comment]
Excellent point. I tried using `abstract` as a marker for the testing framework, and everything worked exactly the way I wanted it to work. Thank you very much for your help.
